### DISCUSSION TITLE
Automations

### DISCUSSION THREAD

**Discussion by veista** [veista/nilan#28] [Q&A]
Should one still be able to run following type of holding register services even your perfect HACS is active? Looking to build Nordpool sensor triggered automations w/o python skills, but e.g. below service does not turn PILP off:

service: modbus.write_register
data:
  address: 1001
  slave: 30
  value: 0
  hub: hub1
  

_Originally posted by @AlvinKleine in https://github.com/veista/nilan/issues/25#issuecomment-1361127150_

**Comment by veista** [comment]
You should create an automation using the integration. This is fairly simple from settings->automations.

**Reply by AlvinKleine** [reply]
Yes thx. Nilan seems not to offer easy way to switch "floor heating" on / off. This would be handy as it's biggest energy consumer. https://hyotynen.iki.fi/kotiautomaatio/tehomaksu/ introduces an option to mimic panel entries which would - on top of many other things - require modbus service triggering.

**Reply by veista** [reply]
I don't think on your modbus version you need such a thing. On bus version 19 most functions are available via modbus. I think this feature might be missing, cause I'm not sure of the correct registers. I believe it is possible to turn central heating off on bus version 19.

**Reply by veista** [reply]
Do you have a sensor called after heating type?

**Reply by AlvinKleine** [reply]
Sorry now I might not catch your Q?

**Reply by veista** [reply]
Yeah, sorry it is available only on modbus 22

**Reply by AlvinKleine** [reply]
registers.py seems to list holding registers for central heating:

    central_heat_heat_extern = 1800
    central_heat_heat_select = 1801
    central_heat_supply_min = 1802
    central_heat_supply_max = 1803
    central_heat_supply_offset = 1804
    central_heat_curve_select = 1805
    central_heat_circ_pump_mode = 1806
    central_heat_heat_type = 1807
    central_heat_reg_time = 1808

**Reply by veista** [reply]
Yes. I need to take a closer look at this when I have time.

**Reply by MKBU2** [reply]
AlvinKleine: What model you have? My device VP 18cek firmware: 2.31.d. Wondering if the external registers could be opened by firmware upgrade?

**Reply by veista** [reply]
Hi, to my understanding yes. It reads in the integration wiki.

**Reply by AlvinKleine** [reply]
> AlvinKleine: What model you have? My device VP 18cek firmware: 2.31.d. Wondering if the external registers could be opened by firmware upgrade?

Model VP18cek, SW 2.38t. I had to renew my broken PCBs (broke pin) but with that latest SW version available.

**Comment by veista** [comment]
Ok, so those registers you mentioned are only available on bus version 22 onwards, excluding reg. 1800 and are already implemented. You can try to read these with qModMaster or eq. if you wish and prove me wrong.

I fixed the diagnostic sensor "After Heating Type" on the latest release. Could you report what this sensor states for you?

**Reply by veista** [reply]
https://www.en.nilan.dk/products/accessories/control-panels/cts602-hmi-upgrade-kit

**Reply by AlvinKleine** [reply]
Is it version 1.1.5B which is fixed?

I actually manage to write to 1801 with following switch:

modbus:
---8>---------
  switches:
  - name: nilan_lammitys_kytkin
    slave: 30
    address: 1801
    write_type: holdings
    command_off: 0
    command_on: 1
    verify:
        input_type: holding
        address: 1801
        state_on: 1
        state_off: 0
        delay: 5
-------------------

....still wondering why I can not do the same with "modbus.write_register", but anyway now Nordpool sensor change triggers automation with some conditions and switches central heating on/off.

**Reply by veista** [reply]
Yes, 1.1.5B. That is great news then. Can you test the other registers too?

**Reply by AlvinKleine** [reply]
> Ok, so those registers you mentioned are only available on bus version 22 onwards, excluding reg. 1800 and are already implemented. You can try to read these with qModMaster or eq. if you wish and prove me wrong.
> 
> I fixed the diagnostic sensor "After Heating Type" on the latest release. Could you report what this sensor states for you?

"No heater".

**Reply by AlvinKleine** [reply]
no luck to write to any registers with service modbus_write.register. E.g.

service: modbus.write_register
data:
  hub: nilan
  address: 1802
  slave: 30
  value: 18

...results in log:  
ERROR (SyncWorker_6) [homeassistant.components.modbus.modbus] Pymodbus: nilan: Exception Response(134, 6, IllegalFunction)

...full log attached:
[modbus_write.register debug.txt](https://github.com/veista/nilan/files/10289955/modbus_write.register.debug.txt)

**Reply by veista** [reply]
Just to clarify, can you test reading holding registers 1801-1808 and also if you could test HR 1209, 1500-1503, 1702-1706, 4015-4019, 4100-4112. Thanks!

**Reply by veista** [reply]
OK, thanks, then there is an error only regarding register 1801 in documentation, I guess. Will fix this for you.

**Reply by AlvinKleine** [reply]
> Just to clarify, can you test reading holding registers 1801-1808 and also if you could test HR 1209, 1500-1503, 1702-1706, 4015-4019, 4100-4112. Thanks!

Read works for all HRs u listed. I suppose there's some problem with "modbus_write.register" in my setup as it does not work for 1801 either. Need to study.

**Reply by veista** [reply]
OK. Can you test does the latest release work for you. Also test Input registers: 5, 1701, 3100, 3102. Holding registers 2002, 2003

**Reply by AlvinKleine** [reply]
Write (with modbus service) still not ok. Read OK for  others than 5.

**Reply by veista** [reply]
Well that is most interesting. Have you gotten a software update to your device by the way?

**Reply by veista** [reply]
I will open up the rest of the registers for your bus version excluding 5, then you will have basically all functionality. Please report if you get modbus errors.

**Reply by veista** [reply]
Does your model still have the enamel boiler with anode?

**Reply by AlvinKleine** [reply]
Yes urethane coated boiler with anode, but I had to renew PCBs in November, so SW should not be too old.

**Reply by veista** [reply]
OK. Apparently Bus Version 19 (FW 2.38) has a lot undocumented registers opened compared to FW 2.37 which is the latest manual to pre Touch Panel. Touch Panel Manual has a change log which gives the impression that they were added only to touch panel versions.

**Reply by AlvinKleine** [reply]
No modbus errors by far, automations work well for setting curve etc. 

Only wonder relates to connection stability. Logbook reports "become unknown" for many sensors every minute or two (see example image attached). 

Any thoughts if this is normal or maybe my setup has some connectivity issues... ...tested by connecting USB-Nano-485 to RPI4 with and without 1m usb cable, same issues.

![image](https://user-images.githubusercontent.com/120299474/209644956-f7d956cd-3e7f-4a40-b27d-fe7bc533387c.png)

**Reply by veista** [reply]
Hi, have noticed this too, especially close to restarts. It might be, now that there are a lot of additions to the integration, the 10 second refresh rate is too high, though this is just a guess since it should be enough time in theory if no errors occur. You can try this by editing the time in const.py

**Reply by AlvinKleine** [reply]
Seems additional config for Modbus integration caused mess. Before founding your connection I had written Modbus config manually. Now after wiping it out issue solved!

**Comment by veista** [comment]
As for the button presses, take a look here #5

**Comment by veista** [comment]
There might be a way to accomplish what you are trying to achieve using the "week program" function. This is not yet implemented, because I haven't seen the need before. You can test it manually if this works for you and I will implement the week program selection  to the integration if it works.

**Comment by veista** [comment]
Added support for Central Heat Select on Bus Version 19 on 1.1.6B

**Reply by AlvinKleine** [reply]
Nice thanks! Changed my automations to use "After Heating Mode option". Reporting results tomorrow.

**Reply by veista** [reply]
Nice. Added remaining entities for Bus Version 19 on 1.1.7B.

**Reply by veista** [reply]
"After Heating Mode" is a typo, should be "Central Heating Mode" will be fixed in next release. Sorry for inconvenience.

**Reply by AlvinKleine** [reply]
Absolute no inconvenience. One "test results" thought: entity nilan.hvac_mode is listed, but info tab says: "This entity is no longer being provided by the nilan integration. If the entity is no longer in use, delete it in settings." Also grayed out on dashboard. Wondering if this is due to latest version update.

**Reply by veista** [reply]
You have to refresh the frontend. Refresh the web page

**Reply by AlvinKleine** [reply]
No help sorry...

![image](https://user-images.githubusercontent.com/120299474/209333251-2957ac1c-cf2b-4b6b-8445-1c6f2fa9e11a.png)

**Reply by AlvinKleine** [reply]
![image](https://user-images.githubusercontent.com/120299474/209333453-7e51c331-3d9b-4eda-8d6c-ed6b1a129edd.png)

**Reply by veista** [reply]
Restart HA?

**Reply by AlvinKleine** [reply]
Embarrassing, but that solved the issue...

**Comment by MKBU2** [comment]
Just to be clear: 
Is it possible to control Central Heating (min/max/curve/etc) in my configuration (VP 18cek / Firmware: 2.31.d)?
Do I need to update firmware to enable additional modbus registers?
@veista were you planning on implementing button based control for Central Heating?

**Reply by veista** [reply]
Hi. To my knowledge NO. You can try the registers with qmodmaster etc. If they work. Button based controls are unreliable implemented with asyncronous commands so unfortunately no.

**Reply by veista** [reply]
> Do I need to update firmware to enable additional modbus registers?

Yes, to my knowledge.

**Comment by arjanmanders** [comment]
Hi! How do I set the fan mode (0, 1, 2, 3 or 4) in a automation? Can't figure out how. Tried set_fan_mode but doesn't work. Changing the speed in the climate card in my dashboard works just fine. Also thanks for the great integration!

**Reply by veista** [reply]
description: ""
mode: single
trigger: []
condition: []
action:
  - service: climate.set_fan_mode
    data:
      fan_mode: "3"
    target:
      device_id: 38a86e15f4bb31f1a44548bfa0484637

**Reply by veista** [reply]
service: climate.set_fan_mode
data:
  fan_mode: "3"
target:
  entity_id: climate.nilan_hvac

**Reply by veista** [reply]
Both work

**Reply by arjanmanders** [reply]
Got it! I think I made a typo at first... ;(

**Reply by holkehan** [reply]
Hey is it possible to set only supply/return fan level and not both at the same time? The set_fan_mode changes both but I would like to only control the supply fan level. Is that possible?

**Reply by veista** [reply]
Yes, but it depends on your modbus version.

**Reply by veista** [reply]
Im not sure, but i think on all models you can set the fan levels in hmi menus. Then you can reserve for example fan level 2 for your custom level.

**Reply by smokker1983b** [reply]
What modbus version are you using? Cant get fan mode working with mine. Thanks

**Reply by veista** [reply]
The integration supports all modbus versions for cts602. What do you mean by fan mode? If you have the old type cts602 panel you need to get a firmware upgrade from nilan to get full functionality.

**Reply by delinend** [reply]
I to is missing the posebility, to controle the Fan speed, via HA and an automation script.
Want to adjust the Fan speed, via a time of day, and via a button in HA.

I have done it before, via a Python script (Domoticz), where I write to Modbus address 1003 (Control.VentSet/Step/0=Off 1-4=Step number)

    def set_userVent(self, speed=2):
        if (speed<0) or (speed>4):
            return
        return self.write_register(1003, speed)

**Reply by veista** [reply]
> I to is missing the posebility, to controle the Fan speed, via HA and an automation script.
> Want to adjust the Fan speed, via a time of day, and via a button in HA.
> 
> I have done it before, via a Python script (Domoticz), where I write to Modbus address 1003 (Control.VentSet/Step/0=Off 1-4=Step number)
> 
>     def set_userVent(self, speed=2):
>         if (speed<0) or (speed>4):
>             return
>         return self.write_register(1003, speed)
> 
> 

https://www.home-assistant.io/docs/automation/basics/

**Reply by delinend** [reply]
Sorry for danish screendumps. But I can't find the "Fan mode" (speed) in automation :-(


![image](https://github.com/veista/nilan/assets/5739273/0f8245d2-16ea-4fa6-8458-039447f73597)


![image](https://github.com/veista/nilan/assets/5739273/4305235e-fb56-4317-be95-e7145d526030)

![image](https://github.com/veista/nilan/assets/5739273/31532cab-c755-47ff-9699-99a80799c0e8)

![image](https://github.com/veista/nilan/assets/5739273/6b8bd6c0-3c66-4469-851c-35d287ff29b3)

![image](https://github.com/veista/nilan/assets/5739273/2fe03abe-57e3-46f7-86df-0fffa7c54a31)

![image](https://github.com/veista/nilan/assets/5739273/90721cb8-aca4-4c05-9d72-9657ba2f0f13)

**Reply by veista** [reply]
I'm not sure what you do not understand from my previous anwser? Call service -> climate.set_fan_mode choose nilan device -> call desired value 1 to 4. I can not make it any more clear. I suggest you read HA documentation, experiment and get to know HA.

**Reply by delinend** [reply]
Thanks for your patience, veista.

You just guided me in the correct direction, with the "Call service -> climate.set_fan_mode choose nilan device ->" line :-)
Now I got it working.

**Comment by maeehart** [comment]
It appears that the button presses discussed in https://github.com/veista/nilan/issues/5 are not available in the current release. Does the code still exist that enables those? Has anybody worked on this? I would love to be able to use the panel in HA.

**Reply by veista** [reply]
No. They are not available anymore. It is technically not possible to add those over wifi. There are other integrations that handle the button presses over hard wire.

If you want to use this integration, I would suggest getting a firmware update from Nilan service.

**Reply by maeehart** [reply]
OK, thanks!
